hail/python/hailtop/batch_client/aioclient.py

```diff
@@ -423,8 +429,13 @@ def create_job(self, image, command, env=None, mount_docker_socket=False,
         self._jobs.append(j)
         return j
 
-    async def _submit_jobs(self, batch_id, byte_job_specs):
-        assert len(byte_job_specs) > 0
+    async def _submit_jobs(self, batch_id, bunch_index):
+        byte_job_specs = self._byte_job_spec_bunches[bunch_index]
+        if byte_job_specs is None:
+            log.info(f'skipping already submitted bunch {bunch_index}')
+            return
```

Comment:
It seems ljike you've made _submit_jobs re-entrant or something?  I think that's unnecessary if you're handing the retries deeper.  This creates unnecessary state.

Reply:
I'm worried about the case of 10M 1kB jobs which is 10GB of data, retransmitting that over (possibly flaky) WiFi only to have batch return 200 OK seems sufficiently bad to warrant the extra state? I guess I'm prematurely optimizing for a case that we don't know for sure will occur.

Reply:
I would fix this a different way.  I don't like bounded_gather for this reason.  I think we need something like AsyncWorkerPool, but that you can wait on at the end.  Then you loop through the specs converting them to bytes and immediately submit at the end of each bunch.  This will sometimes block.  When the bunch submission is complete, the resources for it will be freed.  This is optimal because you only store data for the active requests.  And you get to overlap preparing and submitting the requests.  And it has no state.

Reply:
FYI, I didn't mean this needs to get fixed in this PR (which is blocking Konrad).  I'd just address the reliability and solve the memory issue later.

Reply:
I'm not worried about memory, but network traffic.

Reply:
Oh.  I don't understand the scenario then.  Can you explain in a bit more detail?

Reply:
This was a hold over from the old code, had my wires crossed, it's all gone now.

I agree with the async worker pool. I want that to be fused into pipeline creation actually, because `t.command` is a lot slower than the time taken to transfer said task.